src/hotspot/share/utilities/globalDefinitions.hpp

```diff
@@ -1155,6 +1155,44 @@ inline intx byte_size(void* from, void* to) {
   return (address)to - (address)from;
 }
 
+// Right shifts with signed integers are compiler implementation specific according to the C++ standard.
+// Use a portable version instead.
+inline int64_t shift_right_arithmetic(int64_t value, uint8_t shift_amount) {
+  return value < 0 ? (int64_t)(~(~(uint64_t)value >> shift_amount)) : (int64_t)((uint64_t)value >> shift_amount);
```

Comment:
I believe we have `java_shift_right` already, it assumes signed extension, however.

https://github.com/openjdk/jdk/blob/master/src/hotspot/share/utilities/globalDefinitions.hpp#L1240

Reply:
Yes, I've seen that - was just not too sure about using it as the C++ standard leaves it up to the compiler to decide (most compilers will probably just sign extend). However, I see that we have already been using `java_shift_right`. So it might be fine to just use that as well.

Reply:
We need @kbarrett opinion on C++ code for this case.
I prefer to use already defined functions if they work to avoid duplication.

Reply:
@kimbarrett 

Looks like we've initially pinged the wrong Kim Barrett :-)

Reply:
C++14 5.8/3 In the description of "E1 >> E2" it says "If E1 has a signed type
and a negative value, the resulting value is implementation-defined."

However, C++20 7.6.7/3 further defines integral arithmetic, as part of
requiring two's-complement behavior.
https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0907r3.html
https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p1236r1.html
The corresponding C++20 text is "Right-shift on signed integral types is an
arithmetic right shift, which performs sign-extension."

As discussed in the two's complement proposal, all known modern C++ compilers
already behave that way.  And it is unlikely any would go off and do something
different now, with C++20 tightening things up.

So I think relying on sign extension by right shift is fine.

Reply:
That comment quoted earlier from globalDefinitions.hpp could be expanded to include the above analysis.

Reply:
Thanks a lot Kim for your input and the detailed comments! I've included it as a comment at the `java_shift_right()` method and updated the code to directly use `java_shift_right()` instead of `shift_right_arithmetic()`.